paimon-core/src/main/java/org/apache/paimon/index/PartitionIndex.java

```diff
@@ -80,29 +81,46 @@ public int assign(int hash, IntPredicate bucketFilter) {
             Long number = entry.getValue();
             if (number < targetBucketRowNumber) {
                 entry.setValue(number + 1);
-                hash2Bucket.put(hash, bucket.shortValue());
+                hash2Bucket.put(hash, (short) bucket.intValue());
                 return bucket;
             } else {
                 iterator.remove();
             }
         }
 
-        // 3. create a new bucket
-        for (int i = 0; i < Short.MAX_VALUE; i++) {
-            if (bucketFilter.test(i) && !totalBucket.contains(i)) {
-                hash2Bucket.put(hash, (short) i);
-                nonFullBucketInformation.put(i, 1L);
-                totalBucket.add(i);
-                return i;
+        int maxBucketId =
+                totalBucket.isEmpty()
+                        ? 0
+                        : totalBucket.stream().mapToInt(Integer::intValue).max().getAsInt();
```

Comment:
Here should not invoke the stream every time, you should cache the maxBucketId.

Reply:
Wise consideration :)